src/lib/hash/md5/md5.cpp

```diff
@@ -63,81 +46,90 @@ inline void II(uint32_t& A, uint32_t B, uint32_t C, uint32_t D, uint32_t M)
+
+      GG< 5>(A,B,C,D,M[ 1]+0xF61E2562);   GG< 9>(D,A,B,C,M[ 6]+0xC040B340);
+      GG<14>(C,D,A,B,M[11]+0x265E5A51);   GG<20>(B,C,D,A,M[ 0]+0xE9B6C7AA);
+      GG< 5>(A,B,C,D,M[ 5]+0xD62F105D);   GG< 9>(D,A,B,C,M[10]+0x02441453);
+      GG<14>(C,D,A,B,M[15]+0xD8A1E681);   GG<20>(B,C,D,A,M[ 4]+0xE7D3FBC8);
+      GG< 5>(A,B,C,D,M[ 9]+0x21E1CDE6);   GG< 9>(D,A,B,C,M[14]+0xC33707D6);
+      GG<14>(C,D,A,B,M[ 3]+0xF4D50D87);   GG<20>(B,C,D,A,M[ 8]+0x455A14ED);
+      GG< 5>(A,B,C,D,M[13]+0xA9E3E905);   GG< 9>(D,A,B,C,M[ 2]+0xFCEFA3F8);
+      GG<14>(C,D,A,B,M[ 7]+0x676F02D9);   GG<20>(B,C,D,A,M[12]+0x8D2A4C8A);
+
+      HH< 4>(A,B,C,D,M[ 5]+0xFFFA3942);   HH<11>(D,A,B,C,M[ 8]+0x8771F681);
+      HH<16>(C,D,A,B,M[11]+0x6D9D6122);   HH<23>(B,C,D,A,M[14]+0xFDE5380C);
+      HH< 4>(A,B,C,D,M[ 1]+0xA4BEEA44);   HH<11>(D,A,B,C,M[ 4]+0x4BDECFA9);
+      HH<16>(C,D,A,B,M[ 7]+0xF6BB4B60);   HH<23>(B,C,D,A,M[10]+0xBEBFBC70);
+      HH< 4>(A,B,C,D,M[13]+0x289B7EC6);   HH<11>(D,A,B,C,M[ 0]+0xEAA127FA);
+      HH<16>(C,D,A,B,M[ 3]+0xD4EF3085);   HH<23>(B,C,D,A,M[ 6]+0x04881D05);
+      HH< 4>(A,B,C,D,M[ 9]+0xD9D4D039);   HH<11>(D,A,B,C,M[12]+0xE6DB99E5);
+      HH<16>(C,D,A,B,M[15]+0x1FA27CF8);   HH<23>(B,C,D,A,M[ 2]+0xC4AC5665);
+
+      II< 6>(A,B,C,D,M[ 0]+0xF4292244);   II<10>(D,A,B,C,M[ 7]+0x432AFF97);
+      II<15>(C,D,A,B,M[14]+0xAB9423A7);   II<21>(B,C,D,A,M[ 5]+0xFC93A039);
+      II< 6>(A,B,C,D,M[12]+0x655B59C3);   II<10>(D,A,B,C,M[ 3]+0x8F0CCC92);
+      II<15>(C,D,A,B,M[10]+0xFFEFF47D);   II<21>(B,C,D,A,M[ 1]+0x85845DD1);
+      II< 6>(A,B,C,D,M[ 8]+0x6FA87E4F);   II<10>(D,A,B,C,M[15]+0xFE2CE6E0);
+      II<15>(C,D,A,B,M[ 6]+0xA3014314);   II<21>(B,C,D,A,M[13]+0x4E0811A1);
+      II< 6>(A,B,C,D,M[ 4]+0xF7537E82);   II<10>(D,A,B,C,M[11]+0xBD3AF235);
+      II<15>(C,D,A,B,M[ 2]+0x2AD7D2BB);   II<21>(B,C,D,A,M[ 9]+0xEB86D391);
+
+      A = (digest[0] += A);
+      B = (digest[1] += B);
+      C = (digest[2] += C);
+      D = (digest[3] += D);
+
+      input += 64;
       }
    }
 
-/*
-* Copy out the digest
-*/
-void MD5::copy_out(uint8_t output[])
+void MD5::init(uint32_t digest[4])
    {
-   copy_out_vec_le(output, output_length(), m_digest);
+   const uint32_t MD5_INIT[4] = { 0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476 };
+   copy_mem(digest, MD5_INIT, 4);
+   }
+
+void MD5::add_data(const uint8_t input[], size_t length)
+   {
+   m_md.add_data(input, length);
+   }
+
+void MD5::final_result(uint8_t output[])
+   {
+   m_md.final_result(output);
    }
 
-/*
-* Clear memory of sensitive data
-*/
 void MD5::clear()
    {
-   MDx_HashFunction::clear();
-   zeroise(m_M);
-   m_digest[0] = 0x67452301;
-   m_digest[1] = 0xEFCDAB89;
-   m_digest[2] = 0x98BADCFE;
-   m_digest[3] = 0x10325476;
+   m_md.clear();
+   }
+
+std::unique_ptr<HashFunction> MD5::new_object() const
+   {
+   return std::make_unique<MD5>();
+   }
+
+std::unique_ptr<HashFunction> MD5::copy_state() const
+   {
+   return std::make_unique<MD5>(*this);
    }
```

Comment:
Now that each MDx hash inherits straight from `HashFunction`, these "plumbing methods" are repeated for each class. From what I can see, they all simply forward to their respective `m_md` or create a copy of themselves. Its just copy-and-paste boilerplate, but it does bug me somewhat.

With the full CRTP-approach [I suggested earlier](https://github.com/randombit/botan/pull/3525#discussion_r1184680854) this would go away. At the expense of inheriting from an `MDx` template.